torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
     local_used_work_->wait();
   }
   local_used_maps_reduced_ = false;
+
+  // Rebuilt bucket if this is the first time to rebuilt
+  if (!rebuilt_tensors_.empty()) {
+    rebuildBuckets();
+  }
+}
+
+void Reducer::rebuildBuckets() {
+  TORCH_INTERNAL_ASSERT(
+      rebuilt_tensors_.size() == rebuilt_param_indices_.size(),
+      "rebuild tensor size is not same as rebuild param indices size.");
+  std::vector<std::vector<size_t>> rebuilt_bucket_indices;
+  std::vector<size_t> bucket_size_limits;
+  bucket_size_limits.push_back(DEFAULT_FIRST_BUCKET_BYTES);
+  bucket_size_limits.push_back(bucket_bytes_cap_);
+  rebuilt_bucket_indices = compute_bucket_assignment_by_size(
+      rebuilt_tensors_,
+      bucket_size_limits,
+      expect_sparse_gradients_[0],
+      rebuilt_param_indices_);
+
+  // For rebuilt bucket indices, it needs to be synced across all ranks.
+  // broadcast the newly rebuilt bucket indices from rank 0 in default.
+  // After sync rebuilt bucket indices, initialize buckets for reducer.
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+  auto num_buckets = rebuilt_bucket_indices.size();
+  auto num_elem_per_bucket = rebuilt_bucket_indices[0].size();
+  int64_t total_size = num_buckets * num_elem_per_bucket;
+  at::TensorOptions options;
+  options = options.dtype(at::kInt);
+  options = options.device(replicas_[0][0].device());
+  auto broadcast_tensor = at::empty({total_size}, at::kInt);
+  auto accessor = broadcast_tensor.accessor<int, 1>();
+  for (size_t i = 0; i < num_buckets; i++) {
+    for (size_t j = 0; j < num_elem_per_bucket; j++) {
+      accessor[i * num_elem_per_bucket + j] = rebuilt_bucket_indices[i][j];
+    }
+  }
+  auto broadcast_tensor_device = at::empty({total_size}, options);
+  broadcast_tensor_device.copy_(broadcast_tensor, true);
+  broadcast_coalesced(
+      process_group_, broadcast_tensor_device, broadcast_bucket_size);
+  broadcast_tensor.copy_(broadcast_tensor_device);
+  for (size_t i = 0; i < num_buckets; i++) {
+    for (size_t j = 0; j < num_elem_per_bucket; j++) {
+      rebuilt_bucket_indices[i][j] = accessor[i * num_elem_per_bucket + j];
+    }
+  }
+  was_rebuilt_bucket_ = true;
+  rebuilt_tensors_.clear();
+  rebuilt_param_indices_.clear();
+  mutex_.unlock();
```

Comment:
This is locked in the final callback installed by `mark_variable_ready`, right? Scattering the lock and unlock in two different functions might be error prone. Is there a way that we can apply the lock in a single place?

Reply:
right, it is locked in final callback, that is why I unblock it here, otherwise initialize_buckets will hang, as initialize_buckets is trying to acquire lock to run. Let me think about how to consolidate them

Reply:
had second thought, it is hard to consolidate locks here. initialize_buckets() is a public method that needs lock. it could be called in reducer constructor where no lock before calling, or could be called in rebuild_buckets where we have to unlock it before calling.

Reply:
Unresolve this for now, will come back to this later to discuss if there is a better to way to manage locks.

Reply:
Does the following make sense.

Since `rebuildBuckets` is only called in `finalize_backward`, and `finalize_backward` `mark_variable_ready`, can we move `rebuildBuckets` out of `finalize_backward` and directly call it in the lambda in `mark_variable_ready`? There we have directly control on the lock. Then we can also move `initialize_buckets` out from `rebuildBuckets`, sth like:

```c++
    // in mark_variable_ready
    torch::autograd::Engine::get_default_engine().queue_callback([=] {
      std::lock_guard<std::mutex> lock(this->mutex_);
      this->finalize_backward();
      auto rebuilt_bucket_indices = rebuildBuckets(...);
      lock.unlock();
      initialize_buckets(std::move(rebuilt_bucket_indices));
    });
```

Reply:
sounds good!